File: runtime/js/src/trampoline.ts

export class Delay<T> {
    constructor(public readonly value: () => Promise<Trampoline<T>>) {}
}

export class Result<T> {
    constructor(public readonly value: T) {}
}

export type Trampoline<T> =
    Delay<T> |
    Result<T>;

export async function resolve<T>(trampPromise: Promise<Trampoline<T>>): Promise<Result<T>> {
    let tramp = await trampPromise;
    while(tramp instanceof Delay) {
        tramp = await tramp.value();
    }
1
    return tramp;
}

export function delay<T>(value: () => Promise<Trampoline<T>>): Trampoline<T> {
    return new Delay<T>(value);
}

export function result<T>(value: T): Trampoline<T> {
    return new Result<T>(value);
}

File: backend/jvm/runtime/src/main/java/dev/argon/runtime/Trampoline.java
package dev.argon.runtime;

public sealed interface Trampoline<T> {
	record Result<T>(T value) implements Trampoline<T> {
		public static <T> Result<T> of(T value) {
			return new Result<>(value);
		}
	}

	non-sealed interface Thunk<T> extends Trampoline<T> {
		Trampoline<T> step();
	}

	static <T> T resolve(Trampoline<T> value) {
		while (value instanceof Thunk<T> thunk) {
			value = thunk.step();
		}

		return ((Result<T>) value).value();
	}
}
